### PR TITLE
feat(button): add experimental button

### PR DESCRIPTION
**Pull request by chrsfshr**
Closes #

{{short description}}

#### Changelog

**New**

Added Experimental Button mixin scss -- under feature flag 'components'

**Changed**

wrapped carbon Button scss in 'stable' when feature flag not in use

#### Testing / Reviewing

{{ Add descriptions, steps or a checklist for how reviewers can verify this PR works or not }}

### CODE REVIEW

File: src/globals/scss/_feature-flags.scss [hunk 1]
```diff
@@ -3,7 +3,9 @@
 //-------------------------
 
 // Initialize our feature flag map with default values
-$feature-flags: () !default;
+$feature-flags: (
+  components: true,
```
Comment (joshblack) [inline src/globals/scss/_feature-flags.scss]: Make sure to toggle this off 👍

**Review by asudoh — COMMENTED** [review]
@chrsfshr Great to see the 1st experimental variant! In addition to the default feature flag @joshblack pointed out, there seem various broken variable references causing Travis build failure - Great to see them fixed. Thanks!

**Comment by alisonjoseph** [comment]
This is a good start! 

Looks like this is missing all the experimental mixins, variables. 

Its been a while since I've looked at the duo-components but I think you will need to bring over colors and the theme files also. Look through the files here and compare to what we have in carbon-components https://github.ibm.com/carbon/duo-components/tree/master/src/globals/scss

Should probably copy over the migrate-to-10x docs also.

**Comment by chrsfshr** [comment]
I swapped out security's color vars with carbon's...Not sure if thats the correct solution if not, will try to bring over their theme files as you mentioned @alisonjoseph

**Comment by alisonjoseph** [comment]
When I set the feature flag to true I'm not seeing the new button styles at all? Am I missing something?
screen shot 2018-08-24 at 2 00 17 pm

**Comment by alisonjoseph** [comment]
Looks like you are missing the button styles from the `duo-components` repo. Along with the experimental mixins etc.

File: src/components/button/_button.scss [hunk 2]
```diff
@@ -8,165 +8,247 @@
...
-    &:focus {
-      color: $inverse-01;
+    // Skeleton State
+    .#{$prefix}--btn.#{$prefix}--skeleton {
+      @include skeleton;
+      width: rem(150px);
     }
+  }
+  @include experimental('components') {
+    $button: #{$prefix}--btn;
```
Comment (asudoh) [inline src/components/button/_button.scss]: @IBM/carbon-developers - Do y'all want to see this pattern in all our styles in our experimental blocks? I’m good with this (I see this works for preventing typo), but I don’t want this pattern mixed with hand-writing (within our experimental blocks). Thanks!

**Comment by chrsfshr** [comment]
Just pushing what I had locally not the full updated component yet, need to add in what @alisonjoseph  merged in from her checkbox exp-component

File: src/globals/scss/_feature-flags.scss [hunk 3]
```diff
@@ -3,7 +3,9 @@
 //-------------------------
 
 // Initialize our feature flag map with default values
-$feature-flags: () !default;
+$feature-flags: (
+  components: false,
```
Comment (diego-codes) [inline src/globals/scss/_feature-flags.scss]: Should this flag be named `components-x` to match what is in the button styles?

**Comment by chrsfshr** [comment]
@alisonjoseph or @joshblack when you get a chance take a look I think I've got the exp-button component together. Thanks!

File: src/components/button/_button.scss [hunk 4]
```diff
@@ -171,3 +172,90 @@
...
+
+    &[disabled] {
+      color: $ibm-colors__gray--70;
+    }
+  }
+}
+
+@include exports('button') {
+  @if feature-flag-enabled('components-x') {
+    @include button;
```
Comment (asudoh) [inline src/components/button/_button.scss:256]: Is this line intentional? (Just double-checking if we want `button` mix-in in addition to `button--x`)

**Review by asudoh — COMMENTED** [review]
LGTM basically, just one comment - Great to see this work is getting there!

**Review by alisonjoseph — CHANGES_REQUESTED** [review]
Overall looks good. Will want design to review for sure though.

The colors should all be pulling from the theme tokens, not the color tokens. Also styles should be separated from the original styles, not overwriting them. So remove the line Akira mentioned and have the styles pulled over.

**Comment by chrsfshr** [comment]
Hi @alisonjoseph what do you mean by theme tokens vs color tokens?

**Comment by alisonjoseph** [comment]
I mean the tokens like
`$ui-01` `$brand-02` etc from the `theme.scss` file and specifically the theme--experimental mixin. 

There are also some button variables already defined in there that should have just gotten moved over when copying from the duo repo

> $button-font-weight: 400 !default !global;
>   $button-font-size: 0.875rem !default !global;
>   $button-border-radius: 0 !default !global;
>   $button-height: 48px !default !global;
>   $button-padding: 0.25rem 0.75rem 0 0.75rem !default !global;
>   $button-padding-sm: 0 $spacing-xs !default !global;
>   $button-padding-lg: $spacing-sm !default !global;
>   $button-border-width: 4px !default !global;
>   $button-outline-offset: -5px !default !global;
>   $button-outline: 1px solid $ibm-colors__white !default !global;

**Review by joshblack — CHANGES_REQUESTED** [review]
Could we un-nest selectors and make sure to use the main selector versus the `$button` variables? Only reason is that it can make it harder to search for a selector since we can only use the modifier or descendants in our search.

**Comment by joshblack** [comment]
I don't know if the experimental colors in components can be applied as systematically as our current colors, unfortunately. Might be worth bringing in design here to make sure we have the right color tokens and values, maybe?

**Comment by alisonjoseph** [comment]
Thats true @joshblack I would at least use them where we used them in the `duo-components` repo however.

**Comment by joshblack** [comment]
@alisonjoseph good call, that makes sense.

**Comment by alisonjoseph** [comment]
Can you post staging link for design to review?

**Comment by jeanservaas** [comment]
@chrsfshr Here's button.

Note: Carbon has both a Primary Danger button and a danger button... we have only one danger button... I was going to ask Anna about this, if we need to add a second danger button we can, but in the meantime I wanted to get this to you.

[Button.zip](https://github.com/IBM/carbon-components/files/2407008/Button.zip)

**Comment by alisonjoseph** [comment]
going to close this, I believe @tw15egan is going to start a clean PR since we have updated designs

**Pull request closed without merging**